ui/src/external/monaco.flux.server.ts

```diff
@@ -1,44 +1,118 @@
-import {sendMessage, initialize} from 'src/external/monaco.flux.messages'
+import Deferred from '../utils/Deferred'
+import {
+  LSPResponse,
+  parseResponse,
+  initialize,
+  LSPMessage,
+  completion,
+  didOpen,
+  didChange,
+} from 'src/external/monaco.flux.messages'
+import {registerCompletion} from './monaco.flux.completions'
 
-const queue = []
-let _module = null,
-  loading = false
+import {
+  CompletionItem,
+  CompletionContext,
+  Position,
+} from 'monaco-languageclient/lib/services'
+import {Server} from '@influxdata/flux-lsp-browser'
 
 type BucketCallback = () => Promise<string[]>
 
-export interface LSPServer {
-  send: (string) => Promise<object>
+export interface WASMServer extends Server {
   register_buckets_callback: (BucketCallback) => void
 }
 
-export default function loader(): Promise<LSPServer> {
-  return new Promise(resolve => {
-    if (_module) {
-      resolve(_module as LSPServer)
-      return
-    }
+const queue = []
+let server = null,
+  loading = false
+
+export class LSPServer {
+  private server: WASMServer
+  private messageID: number = 0
+  private buckets: string[] = []
+  private documentVersions: {[key: string]: number} = {}
+
+  constructor(server: WASMServer) {
+    this.server = server
+    this.server.register_buckets_callback(this.getBuckets)
+  }
+
+  getBuckets = () => {
+    return Promise.resolve(this.buckets)
+  }
+
+  updateBuckets(buckets: string[]) {
+    this.buckets = buckets
+  }
+
+  async send(message: LSPMessage): Promise<LSPResponse> {
+    const body = JSON.stringify(message)
+    const fullMessage = `Content-Length: ${body.length}\r\n\r\n${body}`
+    const response = await this.server.process(fullMessage)
 
-    queue.push(resolve)
+    return parseResponse(response)
+  }
 
-    if (loading) {
-      return
+  initialize() {
+    return this.send(initialize(this.currentMessageID))
+  }
+
+  async completionItems(
+    uri: string,
+    position: Position,
+    context: CompletionContext
+  ): Promise<CompletionItem[]> {
+    try {
+      const response = (await this.send(
+        completion(this.currentMessageID, uri, position, context)
```

Comment:
if we're going all OOPy here.. why not define the messages in their methods on the instance? i dont understand the benefit of keeping the functional interface.

Reply:
Readability, also those function construct a type of message but the server has state to pass to them. We don't have to be 100 percent OOP or functional.